### DISCUSSION TITLE
Support for Color array in addition to current CSS color support

### DISCUSSION THREAD

**Discussion by pramilk** [maplibre/maplibre-style-spec#187] [Ideas]
Currently only CSS style (Hex/RGB/RGBA/HSL) colors are supported for a string color input. Color array is supported via expressions. All of these methods are much slower (30-60%)  than a simple RGBA color array. Overhead of expressions makes it even slower than using CSS color string. Currently about 5-8% of Style parsing time is spent on color parsing which is a on a hot code path and is done on both main and worker thread.

Current supported formats:
"red"
"rgb(25 ,0 ,0)"
"#ff3322"
"hsl(120 50% 80%)"

New additional format:
"[25 ,10 , 10]"
"[25 ,10 ,10 , 0.4]"

**If this is something we agree can be supported then I can make the changes.**

Test: https://jsbench.me/84li0k8nkg/1
<img width="1029" alt="image" src="https://github.com/maplibre/maplibre-style-spec/assets/15951646/cc6ae3e5-c017-475b-bb6e-6aa1b004fdde">

**Comment by HarelM** [comment]
This is an interesting idea, overall sounds good to me in order to boost performance.

**Reply by HarelM** [reply]
Cc @louwers

**Reply by pramilk** [reply]
PR: https://github.com/maplibre/maplibre-style-spec/pull/189

**Reply by louwers** [reply]
Please create a tracking issue on maplibre-native if this is merged.

**Comment by kajkal** [comment]
If I may throw in my 2 cents, I believe that css colors are a well-known and recognized system for conveying color information, I think there is no point in introducing a new convention of our own just for performance reasons.
And even if this change were made would it be a recommended format in the documentation? People would stress whether they should convert their `"hsl(0, 100%, 50%)"` to `"[255,0,0,1]"`?
In my opinion, it is better to work on expressions performance rather than to add this kind of color format.

In any case, a good start would be to compare the parsing time of the sample JSON style with colors defined according to the css color specification and the other one where colors are defined in the format proposed here to show the difference on a _realistic_ example. I assume the difference would be a few milliseconds at most - but well, I could be wrong :)

**Reply by HarelM** [reply]
@kajkal thanks for you valuable input. It has some good points.
Would be interesting to discuss this on the TSC meeting.

**Reply by pramilk** [reply]
I think we should not recommend everyone to use this format. Its just another format in which users can provide color. Users who are more performance sensitive can choose to use this format.

Intension here is to extract little bit of performance improvement likely in 10-20ms range (@75th percentile) for styles that we use in Bing maps.  For maplibre default style, if they choose to use this format gains will be lesser (3-10ms) as they have much simpler style. Bing maps currently has 158 Hex and 201 HSL in its style. Maplibre default style has 94 hex, 79 RGB/A, 11 HSL.

Also note that these colors will be parsed on both main and on worker threads.

**Update:**
Here is the screenshot of the profile of Bing maps style. This change reduces CSS parsing time from 21.5 to 4.2+8 on my dev machine. 

Current Bing maps Road style using CSS color format:
![image](https://github.com/maplibre/maplibre-style-spec/assets/15951646/21b19e28-e68d-416e-a30b-5a7a968d4801)

Here is how it looks when using JSON array color for the same style:
![image](https://github.com/maplibre/maplibre-style-spec/assets/15951646/d8941a75-8677-46f4-988e-72da208cebe7)

**BUT**: During loading of the page with Bing maps style (has 372 total layers in its style) even when JSON array color is used, parseCssColor is called hundreds of times for "#000000"  and  "rgba(0, 0, 0, 0)".  This is causing an additional 8ms. 
![image](https://github.com/maplibre/maplibre-style-spec/assets/15951646/d88fbebe-99c4-4aa9-b203-83630b0fc3a9)

Caching these 2 specific colors can save additional 8ms. Will look into it separately.

**Reply by kajkal** [reply]
@pramilk somehow I missed the update to your comment and based mine comment below on the first version I got the notification email about.
I see you also mentioned cache. There is already some [parseColor cache](https://github.com/maplibre/maplibre-style-spec/blob/f1a566141387730a2347956d1ade64173fd58163/src/expression/evaluation_context.ts#L48C5-L54) but from what I can tell it doesn't work very well. 
What I'm suggesting is a cache at the color class level, so something like this:
```ts
class Color {
    static cache = new Map<string, Color>();
    static parse(input: Color | string | undefined | null): Color | undefined {
        // [...]
        const cached = Color.cache.get(input);
        if (cached) {
            return cached;
        }
        const rgba = parseCssColor(input);
        if (rgba) {
            const parsed = new Color(...rgba, false);
            Color.cache.set(input, parsed);
            return parsed;
        }
    }
}
```
It would also be a good idea to review the [default colors](https://github.com/maplibre/maplibre-style-spec/blob/f1a566141387730a2347956d1ade64173fd58163/src/reference/v8.json) and simplify them as much as possible in the order namedColors, hex, rgb, hsl. So for example
- instead of `#ffffff` use `white`
- instead of `rgba(0, 0, 0, 0)` use `transparent`
- instead of `rgba(0, 0, 255, 0)` use `#00f0` etc.

As you said these colors are parsed very often so it wouldn't hurt to make sure that they parse as fast as possible the first time.

**Comment by kajkal** [comment]
This format will duplicate the rgb()/rgba() syntax from css color 3:
`"[25,10,10]"` = `"rgb(25,10,10)"`
`"[25,10,10,0.4]"` = `"rgba(25,10,10,0.4)"`
and I'm not a fan of that.

My thoughts on color parsing performance:
* I suspect that in more complex styles where the issue of performance is more important many of the defined colors are not unique within the scope of a given style, i.e. there are several places where, for example `#fff` is used.
* it might be worth considering implementing a cache for parsed colors so that `#fff` is parsed once and subsequent queries to parse `#fff` return the result of the first parse. `Map<string, Color>`
    (+)  no need to introduce a new color format just for performance reasons -> better legibility of style colors for those unfamiliar with the implementation details of maplibre - recognizable css colors
    (+)  additional benefit for hsl where in case of duplicates a given color is parsed but also converted to rgb only once
    (-) memory leaks; clear cache when a new/other style is requested? cache in scope of single style? color cache size limit as option?
    (+) potential memory leaks not so dangerous, `Color` is basically a small collection of numbers
    (+/-) for simple styles additional memory overhead for color cache, for complex styles some memory savings with cache
* introducing variables, something like in css var(), in the scope of a given style, in one place in the style defined list of variables, then reference them by variable name
    ```
    "vars": {
        "roadColor": "#f00"
    },
    "layers": [{
        "paint": {
            "line-color": "var(--roadColor)", // or ["var", "roadColor"] or "@roadColor" or whatever
        },
    }]
    ````
    (+) as in cache repeating colors are parsed only once
    (+?) way to define style themes, changing colors in one place rather than in each layer
    (+) way to better show dependencies between colors for example bridge and road layers uses the same line color referenced as `roadColor` or not duplicated `hsl(x,y,z)`
    (?) looking through the code once I came across `constants`, a long deprecated feature as I understand which may have had a similar premise. But I'm not sure, if so there must have been some important reasons by which it was deprecated
    (?) `"let"` and `"var"` already exists, so it looks like it would have to be called `constants` :D

@pramilk would you be so kind as to count how many duplicate colors are among these 158 hex and 201 hsl in bing maps?
I checked for mapbox-streets and out of 250 hsl defined there, there are 99 unique (40%) ones so by introducing the cache proposed above I think we could save roughly half of the time spent on color parsing for similarly complex styles. 

maptiler-streets (maplibre default?):
hex: 94, unique: 20 (21%)
rgb: 79, unique: 46 (58%)
hsl: 11, unique: 9 (82%)

**Comment by pramilk** [comment]
Thanks @kajkal. I see 50% improvement with caching and further 50% with using JSON array. I will hold on to current PR https://github.com/maplibre/maplibre-style-spec/pull/189 and will first get the caching done. If there is still need will relook into JSON Array.

**Comment by HarelM** [comment]
The topic was discussed in the TSC meeting today but there were no real inputs in that meeting.
I personally think supporting this format is not a great solution. there was also a question of which one does it use - rgb? hsl etc, so it gets too vague I think...
